### PR TITLE
Fixed TODO: Implement binary search in sparse_matrix.cpp

### PR DESCRIPTION
**Pull request by Sumith1896**
Fixed TODO in sparse_matrix.cpp but this commit seems to break tests. 
There are two more binary search implementation TODOs, if we can fix this so that tests pass we can implement the other two.
cc @thilinarmtb

### CODE REVIEW

**Comment by certik** [comment]
The idea is correct, there is probably some corner case that fails.

**Comment by isuruf** [comment]
You are missing the case where the answer is the `row_start` or `row_end` at the beginning.

**Comment by Sumith1896** [comment]
The corner cases are handled now. The tests still break. What is it that I am doing wrong?

**Comment by thilinarmtb** [comment]
This patch works.

```diff
diff --git a/src/sparse_matrix.cpp b/src/sparse_matrix.cpp
index 878b518..76d901e 100644
--- a/src/sparse_matrix.cpp
+++ b/src/sparse_matrix.cpp
@@ -70,20 +70,23 @@ RCP<const Basic> CSRMatrix::get(unsigned i, unsigned j) const
     unsigned row_end = p_[i + 1];
     unsigned k;
 
-    while (row_start + 1 != row_end) {
+    if (row_start == row_end) {
+        return zero;
+    }
+
+    while (row_start <= row_end) {
         k = (row_start + row_end)/2;
-        if (j_[k] <= j) {
-            row_start = k;
+
+        if (j_[k] == j) {
+            return x_[k];
+        } else if (j_[k] < j){
+            row_start = k + 1;
         } else {
-            row_end = k;
+            row_end = k - 1;
         }
     }
 
-    if (j_[row_start] == j) {
-        return x_[row_start];
-    } else {
-        return zero;
-    }
+    return zero;
 }
 
 void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
```

**Comment by thilinarmtb** [comment]
If a row of `CSRMatrix` is completely empty then `row_start = row_end`, so we need to check that before continuing with the binary search.

**Comment by Sumith1896** [comment]
Thanks @thilinarmtb I didn't know the above fact.
Fixed the commit, tests pass.
Let's wait for Travis.

**Comment by certik** [comment]
Thanks @Sumith1896, @isuruf  and @thilinarmtb for figuring it out. +1 to merge if Travis passes.

**Comment by Sumith1896** [comment]
It's still WIP, there's one last TODO I'm working on.
Let's get this merged afterwards.

**Comment by certik** [comment]
Ok, ping me when it is ready.

File: src/sparse_matrix.cpp [hunk 1]
```diff
@@ -86,10 +94,19 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
 
     unsigned k = p_[i];
     unsigned row_end = p_[i + 1];
+    unsigned mid = k;
 
-    // TODO: Use binary search
-    while (k < row_end && j_[k] < j)
-        k++;
+    while (k < row_end) {
```
Comment (Sumith1896) [inline src/sparse_matrix.cpp]: Wrote the search keeping in mind the post conditions.
The tests fail for this but pass for the other two commits.

Comment (isuruf) [inline src/sparse_matrix.cpp]: In this search if all `j_[k] < j` then `k` is set to `row_end` which was `p_[i+1]` previously. Now it is not. I think that's why the tests fail.

File: src/sparse_matrix.cpp [hunk 2]
```diff
@@ -86,10 +94,23 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
...
+        } else {
+            row_end = mid - 1;
+        }
+    }
 
-    // TODO: Use binary search
-    while (k < row_end && j_[k] < j)
-        k++;
+    if (j_[k] < j) {
+        k = row_end;
```
Comment (Sumith1896) [inline src/sparse_matrix.cpp]: @isuruf This clears your point, doesn't it?

Comment (isuruf) [inline src/sparse_matrix.cpp]: Yes, but setting `k=row_end` introduces several bugs and leaves another bug that I didn't realise.
You need to increase k, (`k++`) and the if condition should check `k < p_[i+1] && j_[k] < j` to avoid out of bounds accessing.

File: src/sparse_matrix.cpp [hunk 3]
```diff
@@ -86,10 +94,23 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
...
+    }
 
     if (neq(e, zero)) {
         if (k < row_end && j_[k] == j) {
```
Comment (isuruf) [inline src/sparse_matrix.cpp]: Here `row_end` should be replaced with `p_[i+1]` since we changed `row_end` in the binary search

File: src/sparse_matrix.cpp (hunk 2 above)
Comment (Sumith1896) [inline src/sparse_matrix.cpp]: I noticed my first clear mistake is `k=row_end` it should have been `k=p_[i+1]`
and linear increment would be no improvement we need to think of something else.

**Comment by Sumith1896** [comment]
Tests pass. Thanks @isuruf.
Now, note that, not for all cases in todo#3 binary search is applied, in some cases it may resort to linear search itself. Point out suggestions, if any.
Else, ready to merge if Travis pass.
cc @certik @thilinarmtb

**Comment by Sumith1896** [comment]
Thanks @thilinarmtb . Your patch looks great.
I tested it in a separate .cpp for different test cases of `k`, `row_end`, `j_[]` and `j`. It gave correct results for all.

**Comment by thilinarmtb** [comment]
Did you append your work to the last commit? Looks like my comment is no more.

**Comment by Sumith1896** [comment]
Oh yeah, since the tests were passing, I rebased and pushed.
My bad, shouldn't have done it in the review stage.
It won't happen from now on.

**Comment by Sumith1896** [comment]
Thilina's comment on the commit which got erased:
What we need to do here is to find a `k` such that, `j_[k - 1] < j <= j_[k]` and 
`p_[i] < k < p_[i + 1]` because the original code looked something like below.

```c++
while (k < row_end && j_[k] < j)
    k++;
```

So, I am not sure whether this is entirely correct. I came up with the following
patch.

```diff
diff --git a/src/sparse_matrix.cpp b/src/sparse_matrix.cpp
index e07a4c3..008c0b0 100644
--- a/src/sparse_matrix.cpp
+++ b/src/sparse_matrix.cpp
@@ -95,24 +95,26 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
     unsigned k = p_[i];
     unsigned row_end = p_[i + 1];
     unsigned end = p_[i + 1];
-    unsigned mid = k;
+    unsigned mid;
  
     while (k < end) {
         mid = (k + end)/2;
-        if (j_[mid] == j) {
+
+        if (mid == k) {
+            if (j_[k] < j) {
+                k++;
+            }
+            break;
+        } else if (j_[mid] >= j && j_[mid - 1] < j) {
             k = mid;
             break;
-        } else if (j_[mid] < j){
-            k = mid + 1;
-        } else {
+        } else if (j_[mid - 1] >= j){
             end = mid - 1;
+        } else {
+            k = mid + 1;
         }
     }
 
-    if (k < p_[i + 1] && j_[k] < j) {
-        k++;
-    }
-
     if (neq(e, zero)) {
         if (k < row_end && j_[k] == j) {
             x_[k] =  e;
```

For a given `mid` value, one of the following cases is true and in the above algorithm
we progress with the binary search accordingly.

1. `mid == k`
2.  `j_[mid - 1] < j <= j_[mid]`
3. `j <= j_[mid - 1]`
4. `j_[mid ] < j`

**Comment by Sumith1896** [comment]
Just for the record, in case we need it in future :smile:

**Comment by thilinarmtb** [comment]
No problem, one can recover it from here: https://github.com/Sumith1896/csympy/commit/7a1833db70cc8b9be69f9ed8f1b34d6f5e901a87

**Comment by thilinarmtb** [comment]
Great ! Now it's easier for others to review.

**Comment by thilinarmtb** [comment]
Please post the link of the commit as well.

**Comment by Sumith1896** [comment]
Binary search todos:
commit for [#1](https://github.com/Sumith1896/csympy/commit/0747d7d3b7697a93341f19cc2c032a596bf35a9a).
commit for [#2](https://github.com/Sumith1896/csympy/commit/dc0d084a91c047567e2aa2984168272752fd3a3c).
commit for [#3](https://github.com/Sumith1896/csympy/commit/896db4b24da6bda39194babf45496c97fc04c2ea).
The commit with @thilinarmtb's comment [here](https://github.com/Sumith1896/csympy/commit/7a1833db70cc8b9be69f9ed8f1b34d6f5e901a87).
@thilinarmtb This is what you were expecting, I hope.

**Comment by thilinarmtb** [comment]
Just the final link would have been enough. Thanks !

**Comment by Sumith1896** [comment]
@certik Ready to merge

**Comment by thilinarmtb** [comment]
@certik , @isuruf : Please have a look at the patch for `CSRMatrix::set`.

File: src/sparse_matrix.cpp [hunk 4]
```diff
@@ -86,10 +94,25 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
...
+        if (mid == k) {
+            if (j_[k] < j) {
+                k++;
+            }
+            break;
+        } else if (j_[mid] >= j && j_[mid - 1] < j) {
+             k = mid;
+             break;
+        } else if (j_[mid - 1] >= j){
+             end = mid - 1;
```
Comment (certik) [inline src/sparse_matrix.cpp]: This and the previous section seems to have inconsistent indentation. Use 4 spaces. You seemed to use 5 sometimes.

**Comment by certik** [comment]
@thilinarmtb what about the patch?

**Comment by Sumith1896** [comment]
I hope the indentation is fine now.

**Comment by Sumith1896** [comment]
Let if know if it is ready, I can squash the last two commits.

File: src/sparse_matrix.cpp [hunk 5]
```diff
@@ -86,10 +94,25 @@ void CSRMatrix::set(unsigned i, unsigned j, const RCP<const Basic> &e)
...
+        mid = (k + end)/2;
+        if (mid == k) {
+            if (j_[k] < j) {
+                k++;
+            }
+            break;
+        } else if (j_[mid] >= j && j_[mid - 1] < j) {
+            k = mid;
+            break;
+        } else if (j_[mid - 1] >= j){
```
Comment (certik) [inline src/sparse_matrix.cpp]: Space before `{`.

**Comment by certik** [comment]
One last little change. 

Otherwise I think it looks good to me. @thilinarmtb what should I look into with regards the patch? Anything that you don't like?

**Comment by Sumith1896** [comment]
@certik Done, shall I squash?

**Comment by certik** [comment]
Yes.

**Comment by Sumith1896** [comment]
Done.

**Comment by thilinarmtb** [comment]
@certik : It would be great if you can double check the algorithm I proposed.

**Comment by certik** [comment]
Ok, I will run this PR and experiment by hand to make sure.

**Comment by certik** [comment]
It looks fine to me and tests pass, so I am merging. If we discover bugs, we can fix them.

**Comment by thilinarmtb** [comment]
Great !!

**Comment by Sumith1896** [comment]
Thanks for merging.
